File: LibGorden/src/main/java/gorden/util/RxCounter.java
package gorden.util;

import java.util.concurrent.TimeUnit;

import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;

/**
 * document Rxjava2.0 计时计数
 * Created by Gordn on 2017/2/27.
 */

public class RxCounter {
    public static Flowable<Long> L(long from, long to) {
        return counter(from, to, 1, TimeUnit.SECONDS);
    }

    public static Flowable<Long> counter(final long from, final long to, int delay, TimeUnit time) {
        return from == to ? Flowable.<Long>empty() :
                Flowable.interval(0, delay, time, AndroidSchedulers.mainThread())
                        .map(new Function<Long, Long>() {
                            @Override
                            public Long apply(Long aLong) throws Exception {
                                if (from > to)
                                    return from - aLong;
                                else
                                    return from + aLong;
                            }
                        }).take(Math.abs(from - to) + 1);
    }

    /**
     * 倒计时
     *
     * @param time 倒计时多少秒
     * @return Flowable持有对象
     */
    public static Flowable<Long> tick(final long time) {
        return time > 0 ? Flowable.interval(0, 1, TimeUnit.SECONDS, AndroidSchedulers.mainThread())
                .map(new Function<Long, Long>() {
                    @Override
                    public Long apply(Long aLong) throws Exception {
                        return time - aLong;
                    }
                }).take(time + 1) : Flowable.<Long>empty();
    }

    public static Flowable<Long> tick(final long time,TimeUnit timeUnit) {
        return time > 0 ? Flowable.interval(0, 1, timeUnit, AndroidSchedulers.mainThread())
                .map(new Function<Long, Long>() {
                    @Override
                    public Long apply(Long aLong) throws Exception {
                        return time - aLong;
                    }
                }).take(time + 1) : Flowable.<Long>empty();
    }
}

File: app/src/main/java/com/cqebd/student/tools/RxCounter.kt
package com.cqebd.student.tools

import io.reactivex.Flowable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Function
import java.util.concurrent.TimeUnit
import kotlin.math.abs

/**
 * 描述
 * Created by gorden on 2018/3/20.
 */
object RxCounter {
    fun counter(from: Long, to: Long): Flowable<Long> {
        return counter(from, to, 1, TimeUnit.SECONDS)
    }

    fun counter(from:Long,to:Long,delay:Int,time:TimeUnit):Flowable<Long>{
        return if (from == to)
            Flowable.empty()
        else
            Flowable.interval(0, delay.toLong(), time, AndroidSchedulers.mainThread())
                    .map {
                        if (from > to)
                            from - it
                        else
                            from + it
                    }.take(abs(from-to)+1)
    }

    /**
     * 倒计时
     *
     * @param time 倒计时多少秒
     * @return Flowable持有对象
     */
    fun tick(time: Long): Flowable<Long> {
        return if (time > 0)
            Flowable.interval(0, 1, TimeUnit.SECONDS, AndroidSchedulers.mainThread())
                    .map {
                        time-it
                    }.take(time+1)
        else
            Flowable.empty()
    }

    fun tick(time: Long, timeUnit: TimeUnit): Flowable<Long> {
        return if (time > 0)
            Flowable.interval(0, 1, timeUnit, AndroidSchedulers.mainThread())
                    .map {
                        time-it
                    }.take(time+1)
        else
            Flowable.empty()
    }
}